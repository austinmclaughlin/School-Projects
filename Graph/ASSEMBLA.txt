git@git.assembla.com:arm2378-cs378-graph.git
