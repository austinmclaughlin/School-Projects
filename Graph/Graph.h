// ----------------------
// projects/graph/Graph.h
// Copyright (C) 2011
// Glenn P. Downing
// ----------------------

#ifndef Graph_h
#define Graph_h

// --------
// includes
// --------

#include <cassert> // assert
#include <cstddef> // size_t
#include <utility> // make_pair, pair
#include <vector> // vector
// -----
// Graph
// -----

class Graph {
public:
	// --------
	// typedefs
	// --------

	typedef unsigned int vertex_descriptor;
	typedef std::pair<vertex_descriptor, vertex_descriptor> edge_descriptor;

	typedef std::vector<vertex_descriptor>::const_iterator vertex_iterator;
	typedef std::vector<edge_descriptor>::const_iterator edge_iterator;
	typedef std::vector<vertex_descriptor>::const_iterator adjacency_iterator;

	typedef std::size_t vertices_size_type;
	typedef std::size_t edges_size_type;

public:
	// --------
	// add_edge
	// --------

	/**
	 * @param vd1 the source vertex for this edge
	 * @param vd2 the target vertex for this edge
	 * @param graph the graph vd1 and vd2 belong to
	 * @return a pair containing the edge_descriptor as its first element and a bool as its second element
	 * the bool is true if this edge was just added
	 * @throws invalid_argument if vertex_descriptor not valid
	 * @throws invalid_argument if edge exists
	 */
	friend std::pair<edge_descriptor, bool> add_edge(vertex_descriptor vd1,
			vertex_descriptor vd2, Graph& graph) {

		std::pair<edge_descriptor, bool> p = edge(vd1, vd2, graph);
		edge_descriptor ed;
		bool b;
		if (p.second) {
			ed = p.first;
			b = false;
		} else {
			graph.g[vd1].push_back(vd2);
			ed = std::make_pair(vd1, vd2);
			graph.elist.push_back(ed);
			b = true;
		}
		return std::make_pair(ed, b);
	}

	// ----------
	// add_vertex
	// ----------

	/**
	 * @param graph is the graph a new vertex will be added to
	 * @return a vertex_descriptor for the added vertex
	 */
	friend vertex_descriptor add_vertex(Graph& graph) {
		vertex_descriptor vd = graph.g.size();
		graph.vlist.push_back(vd);
		std::vector<vertex_descriptor> v;
		graph.g.push_back(v);
		return vd;
	}

	// -----------------
	// adjacent_vertices
	// -----------------

	/**
	 * @param vd a vertex_descriptor in graph whose adjacent vertices are requested
	 * @param graph the graph vd belongs to
	 * @return a pair containing an iterator to the beginning and end of the adjacent vertices of vd
	 * @throws invalid_argument if vertex_descriptor not valid
	 */
	friend std::pair<adjacency_iterator, adjacency_iterator> adjacent_vertices(
			vertex_descriptor vd, const Graph& graph) {
		adjacency_iterator b = graph.g[vd].begin();
		adjacency_iterator e = graph.g[vd].end();
		return std::make_pair(b, e);
	}

	// ----
	// edge
	// ----

	/**
	 * @param vd1 the source vertex for this edge
	 * @param vd2 the target vertex for this edge
	 * @param graph the graph containing vd1 and vd2
	 * @return a pair containing an edge_descriptor between vd1 and vd2 and a bool to signify if the edge was added
	 * @throws invalid_argument if vertex_descriptor not valid
	 */
	friend std::pair<edge_descriptor, bool> edge(vertex_descriptor vd1,
			vertex_descriptor vd2, const Graph& graph) {
		edge_descriptor ed = std::make_pair(0, 0);
		bool b = false;
		if (vd1 < graph.g.size()) {
			if (vd2 < graph.g[vd1].size()) {
				ed = std::make_pair(vd1, vd2);
				b = true;
			}
		}
		return std::make_pair(ed, b);
	}

	// -----
	// edges
	// -----

	/**
	 * O(1) in space
	 * O(1) in time
	 * <your documentation>
	 */
	friend std::pair<edge_iterator, edge_iterator> edges(const Graph& graph) {
		// <your code>
		edge_iterator b = graph.elist.begin();
		edge_iterator e = graph.elist.end();
		return std::make_pair(b, e);
	}

	// ---------
	// num_edges
	// ---------

	/**
	 * @param graph the graph in question
	 * @return the number of edges in graph
	 */
	friend edges_size_type num_edges(const Graph& graph) {
		return graph.elist.size();
	}

	// ------------
	// num_vertices
	// ------------

	/**
	 * @param graph the graph in question
	 * @return the number of vertices in graph
	 */
	friend vertices_size_type num_vertices(const Graph& graph) {
		return graph.vlist.size();
	}

	// ------
	// source
	// ------

	/**
	 * @param ed an edge_descriptor
	 * @param graph the graph ed belongs to
	 * @return the source of ed
	 * @throws invalid_argument if edge_descriptor not valid
	 */
	friend vertex_descriptor source(edge_descriptor ed, const Graph& graph) {
		assert (ed.first < graph.g.size());
		return ed.first;
	}

	// ------
	// target
	// ------

	/**
	 * @param ed an edge_descriptor
	 * @param graph the graph ed belongs to
	 * @return the target of ed
	 * @throws invalid_argument if edge_descriptor not valid
	 */
	friend vertex_descriptor target(edge_descriptor ed, const Graph& graph) {
		assert (ed.first < graph.g.size());
		assert (ed.second < graph.g.size());
		return ed.second;
	}

	// ------
	// vertex
	// ------

	/**
	 * @param v
	 * @param graph the graph v belongs to
	 * @throws invalid_argument if vertices_size_type not valid
	 */
	friend vertex_descriptor vertex(vertices_size_type v, const Graph& graph) {
		assert (v < graph.g.size());
		return v;
	}

	// --------
	// vertices
	// --------

	/**
	 * @param graph
	 * @return a pair of iterators to the beginning and end of the vertices in graph
	 */
	friend std::pair<vertex_iterator, vertex_iterator> vertices(
			const Graph& graph) {
		vertex_iterator b = graph.vlist.begin();
		vertex_iterator e = graph.vlist.end();
		return std::make_pair(b, e);
	}

private:
	// ----
	// data
	// ----

	std::vector<vertex_descriptor> vlist;
	std::vector<edge_descriptor> elist;
	std::vector<std::vector<vertex_descriptor> > g; // something like this

	// -----
	// valid
	// -----

	/**
	 * <your documentation>
	 */
	bool valid() const {
		bool val = true;
		if (vlist.size() < 0 || elist.size() < 0)
			val = false;
		return val;
	}

public:
	// ------------
	// constructors
	// ------------

	/**
	 * <your documentation>
	 */
	Graph() {
		assert(valid());
	}

	// Default copy, destructor, and copy assignment
	// Graph (const Graph<T>&);
	// ~Graph ();
	// Graph& operator = (const Graph&);
};

// ---------
// has_cycle
// ---------

/**
 * depth-first traversal
 * three colors
 *
 * @param g is the graph to check
 * @return bool true if g has a cycle, false otherwise
 */
template<typename G>
bool has_cycle(const G& g) {

	typedef typename G::vertex_descriptor vertex_descriptor;
	typedef typename G::vertex_iterator vertex_iterator;

	std::set<vertex_descriptor> grey_list;
	std::set<vertex_descriptor> black_list;

	std::pair<vertex_iterator, vertex_iterator> p = vertices(g);
	vertex_iterator b = p.first;
	vertex_iterator e = p.second;

	bool r = false;
	while (b != e && !r) {
		if (black_list.find(*b) == black_list.end())
			r = has_cycle_helper(*b, grey_list, black_list, g);
		++b;
	}
	return r;
}

/**
 * helper method for has_cycle
 * recursively checks node vd and its children for a grey node
 * if a grey node is found, we have a cycle and has_cycle_helper returns true
 *
 * @param vd is the vertex to check
 * @param grey_list is a reference to the list of grey nodes
 * @param black_list is a reference to the list of black nodes
 * @return true if vd or one of its children is grey, false otherwise
 */
template<typename G>
bool has_cycle_helper(typename G::vertex_descriptor vd,
		std::set<typename G::vertex_descriptor>& grey_list,
		std::set<typename G::vertex_descriptor>& black_list, const G& g) {

	std::pair<typename G::adjacency_iterator, typename G::adjacency_iterator>
			p = adjacent_vertices(vd, g);
	typename G::adjacency_iterator b = p.first;
	typename G::adjacency_iterator e = p.second;

	bool r = false;
	if (grey_list.find(vd) == grey_list.end()) {
		grey_list.insert(vd);
		while (b != e && !r) {
			if (black_list.find(*b) == black_list.end())
				r = has_cycle_helper(*b, grey_list, black_list, g);
			++b;
		}
		if (!r) {
			grey_list.erase(vd);
			black_list.insert(vd);
		}
	} else {
		r = true;
	}

	return r;
}

// ----------------
// topological_sort
// ----------------

/**
 * depth-first traversal
 * calls tsort_helper with all nodes that have no incoming vertices
 * @param g is the graph to check
 * @param x is the output iterator to write the sorted results to
 * @throws invalid_argument if !has_cycle(g)
 */
template<typename G, typename OI>
void topological_sort(const G& g, OI x) {

	typedef typename G::vertex_descriptor vertex_descriptor;
	typedef typename G::vertex_iterator vertex_iterator;
	typedef typename G::edge_descriptor edge_descriptor;
	typedef typename G::edge_iterator edge_iterator;

	assert(!has_cycle(g));

	std::vector<vertex_descriptor> sorted; //list that will contain the sorted nodes (will be in reverse order)
	std::set<vertex_descriptor> roots; //all nodes with no incoming edges
	std::set<vertex_descriptor> visited;

	std::pair<vertex_iterator, vertex_iterator> p1 =
			vertices(g);
	std::pair<edge_iterator, edge_iterator> p2 = edges(
			g);
	vertex_iterator b1 = p1.first;
	vertex_iterator e1 = p1.second;
	edge_iterator b2 = p2.first;
	edge_iterator e2 = p2.second;

	while (b1 != e1) {
		roots.insert(*b1);
		++b1;
	}
	while (b2 != e2) {
		roots.erase(target(*b2, g));
		++b2;
	}

	typename std::set<vertex_descriptor, std::less<vertex_descriptor>,
			std::allocator<vertex_descriptor> >::iterator b3 = roots.begin();
	typename std::set<vertex_descriptor, std::less<vertex_descriptor>,
			std::allocator<vertex_descriptor> >::iterator e3 = roots.end();

	while (b3 != e3) {
		tsort_helper(*b3, sorted, visited, g);
		++b3;
	}

	while (!sorted.empty()) {
		*x = sorted.back();
		sorted.pop_back();
		++x;
	}
}

/**
 * checks individual vertices
 * if a vertex has been visited, do nothing
 * otherwise, mark the vertex as visited, and recursively call tsort_helper on each
 * vertex this vd points to.
 * after all child vertices have been visited, add vd to the sorted list
 * @param vd is the vertex to check
 * @param sorted is a vector of vertices in reverse sorted order
 * @param visited is the list of vertices that have been visited
 * @param g is the graph vd belongs to
 */
template<typename G>
void tsort_helper(typename G::vertex_descriptor vd,
		std::vector<typename G::vertex_descriptor>& sorted,
		std::set<typename G::vertex_descriptor>& visited, const G& g) {

	if (visited.find(vd) == visited.end()) {
		visited.insert(vd);

		std::pair<typename G::adjacency_iterator,
				typename G::adjacency_iterator> p = adjacent_vertices(vd, g);
		typename G::adjacency_iterator b = p.first;
		typename G::adjacency_iterator e = p.second;

		while (b != e) {
			tsort_helper(*b, sorted, visited, g);
			++b;
		}
		sorted.push_back(vd);
	}
}

#endif // Graph_h
