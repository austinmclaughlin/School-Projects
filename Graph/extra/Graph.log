commit 4a242aed29a158c1c33967f92d5000a2ad7d3ff6
Author: Austin McLaughlin <txaustinm@gmail.com>
Date:   Sat Jul 30 01:05:15 2011 -0500

    fixed bug in tsort_helper and added tests for topological_sort

commit 1181f2ba06c0437b4c091622506738e5fec98f76
Author: Austin McLaughlin <txaustinm@gmail.com>
Date:   Fri Jul 29 18:25:13 2011 -0500

    implemented topological_sort

commit 04aa909b89ee9fb5e95a88cd6f7a82412bd23cce
Author: Austin McLaughlin <txaustinm@gmail.com>
Date:   Fri Jul 29 01:35:17 2011 -0500

    implemented has_cycle, seems to be working on current tests

commit 402ec51382d82b598e4fbeb57bfdf75cda0fd76f
Author: Austin McLaughlin <txaustinm@gmail.com>
Date:   Wed Jul 27 00:32:06 2011 -0500

    added documentation

commit 9ed654af3873ac41d93f1bde2f0a1b35da541a16
Author: Austin McLaughlin <txaustinm@gmail.com>
Date:   Sun Jul 24 16:56:16 2011 -0500

    implemented constructor, add_vertex, add_edge, edge, num_edges, num_vertices, source, and targe

commit ba4a1978e5d5c00d33bab03b5de2e102aa8750e9
Author: Austin McLaughlin <txaustinm@gmail.com>
Date:   Sat Jul 23 14:04:38 2011 -0500

    previously uploaded source files were incorrect, these are the right ones

commit 334a242b30870fcd4fb6a8430f411e1010c3d10e
Author: Austin McLaughlin <txaustinm@gmail.com>
Date:   Sat Jul 23 13:54:28 2011 -0500

    added source files

commit 8de1263291630f2f2d2f10c962a3862f18452fdd
Author: Austin McLaughlin <txaustinm@gmail.com>
Date:   Sat Jul 23 13:52:57 2011 -0500

    Initialize repository

