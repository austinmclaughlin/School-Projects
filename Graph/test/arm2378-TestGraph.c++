// ----------------------------
// projects/graph/TestGraph.c++
// Copyright (C) 2011
// Glenn P. Downing
// ----------------------------

/*
 To test the program:
 % g++ -ansi -pedantic -I/public/linux/include/boost-1_44 -lcppunit -ldl -Wall TestGraph.c++ -o TestGraph.app
 % valgrind TestGraph.app >& TestGraph.c++.out
 */

// --------
// includes
// --------

#include <iostream> // cout, endl
#include <iterator> // ostream_iterator
#include <sstream> // ostringstream
#include <utility> // pair
#include "boost/graph/adjacency_list.hpp" // adjacency_list
#include "cppunit/extensions/HelperMacros.h" // CPPUNIT_TEST, CPPUNIT_TEST_SUITE, CPPUNIT_TEST_SUITE_END
#include "cppunit/TestFixture.h" // TestFixture
#include "cppunit/TestSuite.h" // TestSuite
#include "cppunit/TextTestRunner.h" // TestRunner
#include "Graph.h"

// ---------
// TestGraph
// ---------

template<typename T>
struct TestGraph: CppUnit::TestFixture {
	// --------
	// typedefs
	// --------

	typedef T graph_type;

	typedef typename graph_type::vertex_descriptor vertex_descriptor;
	typedef typename graph_type::edge_descriptor edge_descriptor;

	typedef typename graph_type::vertex_iterator vertex_iterator;
	typedef typename graph_type::edge_iterator edge_iterator;
	typedef typename graph_type::adjacency_iterator adjacency_iterator;

	typedef typename graph_type::vertices_size_type vertices_size_type;
	typedef typename graph_type::edges_size_type edges_size_type;

	// -----
	// tests
	// -----

	// g
	graph_type g;

	vertex_descriptor vdA;
	vertex_descriptor vdB;
	vertex_descriptor vdC;
	vertex_descriptor vdD;
	vertex_descriptor vdE;
	vertex_descriptor vdF;
	vertex_descriptor vdG;
	vertex_descriptor vdH;

	edge_descriptor edAB;
	edge_descriptor edAC;
	edge_descriptor edAE;
	edge_descriptor edBD;
	edge_descriptor edBE;
	edge_descriptor edCD;
	edge_descriptor edDE;
	edge_descriptor edDF;
	edge_descriptor edFD;
	edge_descriptor edFH;
	edge_descriptor edGH;

	// g2
	graph_type g2;

	vertex_descriptor vdL;
	vertex_descriptor vdM;
	vertex_descriptor vdN;
	vertex_descriptor vdO;
	vertex_descriptor vdP;

	edge_descriptor edLM;
	edge_descriptor edLN;
	edge_descriptor edOM;
	edge_descriptor edOP;

	// g3
	graph_type g3;

	vertex_descriptor vdX;
	vertex_descriptor vdY;
	vertex_descriptor vdZ;

	edge_descriptor edXY;
	edge_descriptor edXZ;

	// -----
	// setUp
	// -----

	// directed, sparse, unweighted
	// possibly connected
	// possibly cyclic
	// Collins, 2nd, pg. 653
	void setUp() {
		vdA = add_vertex(g);
		vdB = add_vertex(g);
		vdC = add_vertex(g);
		vdD = add_vertex(g);
		vdE = add_vertex(g);
		vdF = add_vertex(g);
		vdG = add_vertex(g);
		vdH = add_vertex(g);
		edAB = add_edge(vdA, vdB, g).first;
		edAC = add_edge(vdA, vdC, g).first;
		edAE = add_edge(vdA, vdE, g).first;
		edBD = add_edge(vdB, vdD, g).first;
		edBE = add_edge(vdB, vdE, g).first;
		edCD = add_edge(vdC, vdD, g).first;
		edDE = add_edge(vdD, vdE, g).first;
		edDF = add_edge(vdD, vdF, g).first;
		edFD = add_edge(vdF, vdD, g).first;
		edFH = add_edge(vdF, vdH, g).first;
		edGH = add_edge(vdG, vdH, g).first;

		vdL = add_vertex(g2);
		vdM = add_vertex(g2);
		vdN = add_vertex(g2);
		vdO = add_vertex(g2);
		vdP = add_vertex(g2);
		edLM = add_edge(vdL, vdM, g2).first;
		edLN = add_edge(vdL, vdN, g2).first;
		edOM = add_edge(vdO, vdM, g2).first;
		edOP = add_edge(vdO, vdP, g2).first;

		vdX = add_vertex(g3);
		vdY = add_vertex(g3);
		vdZ = add_vertex(g3);
		edXY = add_edge(vdX, vdY, g).first;
		edXZ = add_edge(vdX, vdZ, g).first;

	}

	// -------------
	// test_add_edge
	// -------------

	void test_add_edge_1() {
		std::pair<edge_descriptor, bool> p = add_edge(vdA, vdB, g);
		CPPUNIT_ASSERT(p.first == edAB);
		CPPUNIT_ASSERT(p.second == false);
	}

	void test_add_edge_2() {
		std::pair<edge_descriptor, bool> p = add_edge(vdA, vdY, g3); // ??? does it auto add the vertex?
		//CPPUNIT_ASSERT(p.first == ed); should be (0,0)
		CPPUNIT_ASSERT(p.second == true);
	}

	// ----------------------
	// test_adjacent_vertices
	// ----------------------

	void test_adjacent_vertices_1() {
		std::pair<adjacency_iterator, adjacency_iterator> p =
				adjacent_vertices(vdA, g);
		adjacency_iterator b = p.first;
		adjacency_iterator e = p.second;
		CPPUNIT_ASSERT(b != e);
		if (b != e) {
			vertex_descriptor vd = *b;
			CPPUNIT_ASSERT(vd == vdB);
		}
		++b;
		if (b != e) {
			vertex_descriptor vd = *b;
			CPPUNIT_ASSERT(vd == vdC);
		}
		++b;
		if (b != e) {
			vertex_descriptor vd = *b;
			CPPUNIT_ASSERT(vd == vdE);
		}
		++b;
		CPPUNIT_ASSERT(b == e);
	}

	// ---------
	// test_edge
	// ---------

	void test_edge_1() {
		std::pair<edge_descriptor, bool> p = edge(vdA, vdB, g);
		CPPUNIT_ASSERT(p.first == edAB);
		CPPUNIT_ASSERT(p.second == true);
		p = edge(vdA, vdF, g);
		edge_descriptor edge;
		CPPUNIT_ASSERT(p.first == edge);
		CPPUNIT_ASSERT(p.second == false);
	}

	// ----------
	// test_edges
	// ----------

	void test_edges_1() {
		std::pair<edge_iterator, edge_iterator> p = edges(g);
		edge_iterator b = p.first;
		edge_iterator e = p.second;
		CPPUNIT_ASSERT(b != e);
		if (b != e) {
			edge_descriptor ed = *b;
			CPPUNIT_ASSERT(ed == edAB);
		}
		++b;
		if (b != e) {
			edge_descriptor ed = *b;
			CPPUNIT_ASSERT(ed == edAC);
		}
		++b;
		if (b != e) {
			edge_descriptor ed = *b;
			CPPUNIT_ASSERT(ed == edAE);
		}
		++b;
		if (b != e) {
			edge_descriptor ed = *b;
			CPPUNIT_ASSERT(ed == edBD);
		};
	}

	// --------------
	// test_num_edges
	// --------------

	void test_num_edges_1() {
		edges_size_type es = num_edges(g);
		CPPUNIT_ASSERT(es == 11);
	}

	// -----------------
	// test_num_vertices
	// -----------------

	void test_num_vertices_1() {
		vertices_size_type vs = num_vertices(g);
		CPPUNIT_ASSERT(vs == 8);
	}

	// -----------
	// test_source
	// -----------

	void test_source_1() {
		vertex_descriptor vd = source(edAB, g);
		CPPUNIT_ASSERT(vd == vdA);
	}

	// -----------
	// test_target
	// -----------

	void test_target_1() {
		vertex_descriptor vd = target(edAB, g);
		CPPUNIT_ASSERT(vd == vdB);
	}

	// -----------
	// test_vertex
	// -----------

	void test_vertex_1() {
		vertex_descriptor vd = vertex(0, g);
		CPPUNIT_ASSERT(vd == vdA);
	}

	// -------------
	// test_vertices
	// -------------

	void test_vertices_1() {
		std::pair<vertex_iterator, vertex_iterator> p = vertices(g);
		vertex_iterator b = p.first;
		vertex_iterator e = p.second;
		CPPUNIT_ASSERT(b != e);
		if (b != e) {
			vertex_descriptor vd = *b;
			CPPUNIT_ASSERT(vd == vdA);
		}
		++b;
		if (b != e) {
			vertex_descriptor vd = *b;
			CPPUNIT_ASSERT(vd == vdB);
		}
	}

	// --------------
	// test_has_cycle
	// --------------

	void test_has_cycle_1() {
		CPPUNIT_ASSERT(has_cycle(g));
	}

	void test_has_cycle_2() {
		CPPUNIT_ASSERT(!has_cycle(g2));
	}

	void test_has_cycle_3() {
		CPPUNIT_ASSERT(!has_cycle(g3));
	}

	// ---------------------
	// test_topological_sort
	// ---------------------

	void test_topological_sort_1() {
		std::ostringstream out;
		topological_sort(g2, std::ostream_iterator<vertex_descriptor>(out, " "));
		CPPUNIT_ASSERT(out.str() == "3 4 0 2 1 ");
	}

	void test_topological_sort_2() {
		std::ostringstream out;
		topological_sort(g3, std::ostream_iterator<vertex_descriptor>(out, " "));
		CPPUNIT_ASSERT(out.str() == "2 1 0 ");
	}

	// -----
	// suite
	// -----
CPPUNIT_TEST_SUITE(TestGraph);
		CPPUNIT_TEST(test_add_edge_1);
		CPPUNIT_TEST(test_add_edge_2);
		CPPUNIT_TEST(test_adjacent_vertices_1);
		CPPUNIT_TEST(test_edge_1);
		CPPUNIT_TEST(test_edges_1);
		CPPUNIT_TEST(test_num_edges_1);
		CPPUNIT_TEST(test_num_vertices_1);
		CPPUNIT_TEST(test_source_1);
		CPPUNIT_TEST(test_target_1);
		CPPUNIT_TEST(test_vertex_1);
		CPPUNIT_TEST(test_vertices_1);
		CPPUNIT_TEST(test_has_cycle_1);
		CPPUNIT_TEST(test_has_cycle_2);
		CPPUNIT_TEST(test_has_cycle_3);
		CPPUNIT_TEST(test_topological_sort_1);
		CPPUNIT_TEST(test_topological_sort_2);
	CPPUNIT_TEST_SUITE_END();
};

// ----
// main
// ----

int main() {
	using namespace std;
	using namespace boost;

	ios_base::sync_with_stdio(false); // turn off synchronization with C I/O
	cout << "TestGraph.c++" << endl;

	CppUnit::TextTestRunner tr;
	tr.addTest(TestGraph<adjacency_list<setS, vecS, directedS> >::suite());
	//tr.addTest(TestGraph<Graph>::suite()); // uncomment
	tr.run();

	cout << "Done." << endl;
	return 0;
}

