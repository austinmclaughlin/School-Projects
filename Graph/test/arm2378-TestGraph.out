==11818== Memcheck, a memory error detector
==11818== Copyright (C) 2002-2010, and GNU GPL'd, by Julian Seward et al.
==11818== Using Valgrind-3.6.1 and LibVEX; rerun with -h for copyright info
==11818== Command: ./TestGraph.c++.app
==11818== 
TestGraph.c++
................


OK (16 tests)


Done.
==11818== 
==11818== HEAP SUMMARY:
==11818==     in use at exit: 122,880 bytes in 6 blocks
==11818==   total heap usage: 1,280 allocs, 1,274 frees, 248,001 bytes allocated
==11818== 
==11818== LEAK SUMMARY:
==11818==    definitely lost: 0 bytes in 0 blocks
==11818==    indirectly lost: 0 bytes in 0 blocks
==11818==      possibly lost: 0 bytes in 0 blocks
==11818==    still reachable: 122,880 bytes in 6 blocks
==11818==         suppressed: 0 bytes in 0 blocks
==11818== Rerun with --leak-check=full to see details of leaked memory
==11818== 
==11818== For counts of detected and suppressed errors, rerun with: -v
==11818== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 6 from 6)
